packages/babel-parser/src/util/expression-scope.js

```diff
@@ -130,6 +131,41 @@ export default class ExpressionScopeHandler {
     /* eslint-disable @babel/development-internal/dry-error-messages */
     this.raise(pos, message);
   }
+
+  /**
+   * Record parenthesized identifier errors
+   *
+   * A parenthesized identifier in LHS can be ambiguous because the assignment
+   * can be transformed to an assignable later, but not vice versa:
+   * For example, in `([(a) = []] = []) => {}`, we think `(a) = []` is an LHS in `[(a) = []]`,
+   * an LHS within `[(a) = []] = []`. However the LHS chain is then transformed by toAssignable,
+   * and we should throw assignment `(a)`, which is only valid in LHS. Hence we record the
+   * location of parenthesized `(a)` to current scope if it is one of MaybeArrowParameterDeclaration
+   * and MaybeAsyncArrowParameterDeclaration
+   *
+   * Unlike `recordParameterInitializerError`, we don't record to ancestry scope because we
+   * validate arrow head parsing scope before exit, and then the LHS will be unambiguous:
+   * For example, in `( x = ( [(a) = []] = [] ) ) => {}`, we should not record `(a)` in `( x = ... ) =>`
+   * arrow scope because when we finish parsing `( [(a) = []] = [] )`, it is an umbiguous assignment
```

Comment:
Nit:

```suggestion
   * arrow scope because when we finish parsing `( [(a) = []] = [] )`, it is an unambiguous assignment
```

(btw, as @nicolo-ribaudo mentioned, <3 <3 <3 these comments)